### ISSUE TITLE
Add testing submodule

### ISSUE THREAD

**Issue by pganssle** [dateutil/dateutil#438]
I have a few little testing utilities in [dateutil/test/_common.py](https://github.com/dateutil/dateutil/blob/master/dateutil/test/_common.py). I think that maybe some of them should be made permanent, for example [`TzContextBase`](https://github.com/dateutil/dateutil/blob/master/dateutil/test/_common.py#L95) and its derivatives could be used. Might be useful for people who want to test their stuff under different time zones.

That said, the way I'm doing it (especially on Windows) involves actually changing the system time zone, which may be a bit overkill compared to just mocking out all the relevant parts that return time zone information. Still, it's a start.

**Comment by jbrockmendel** [comment]
Is the idea to expose this via dateutil/exposed_name.py instead of keeping it private in dateutil/test/?  i.e. is it intended to be used downstream?

**Comment by pganssle** [comment]
@jbrockmendel Yes, similar to [numpy.testing](https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.testing.html) and [pandas.testing](https://pandas.pydata.org/pandas-docs/stable/api.html#testing-functions), the idea would be to add a `dateutil.testing` module that would provide datetime-related functions for testing and/or mocking, starting with context managers (which could maybe be converted into [`contextlib.ContextDecorator`](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator)-type objects.

Another possibility would be to spin these off into their own library, or see if they would be welcome in [`freezegun`](https://github.com/spulec/freezegun). CC @spulec thoughts?

It may be that there's a better way to handle them, though. The Windows one especially is a bit dangerous because it changes the time zone in the OS globally, not just in the local environment.
